fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,351 @@
+                )
+                .await
+                .context("timeout while fetch source"),
+            )
+        }))
+        .await;
+        // initially write all once
+        let mut dbtx = client.db().begin_transaction().await;
+        for (source_id, meta) in metas {
+            if let Ok(meta) = meta.and_then(|x| x) {
+                Self::update_meta_values(&mut dbtx, source_id, &meta).await;
+            }
+        }
+        dbtx.commit_tx().await;
+        self.initial_fetch_waiter.done();
+
+        // now keep updating all
+        let values = join_all(
+            self.sources
+                .iter()
+                .map(|source| self.update_source_continuously(client, &**source)),
+        )
+        .await;
+        // to get ! type
+        values[0]
+    }
+
+    async fn update_source_continuously(&self, client: &Client, source: &dyn MetaSource) -> ! {
+        loop {
+            MetaSource::wait_for_update(source).await;
+            if let Ok(meta) = retry(
+                "update_source",
+                // very slow retry
+                FibonacciBackoff::default()
+                    .with_min_delay(Duration::from_secs(10))
+                    .with_max_delay(Duration::from_secs(10 * 60))
+                    .with_max_times(usize::MAX),
+                || source.fetch(client),
+            )
+            .await
+            {
+                let mut dbtx = client.db().begin_transaction().await;
+                Self::update_meta_values(&mut dbtx, source.source_id(), &meta).await;
+                // TODO: what should have happen in case of conflicts?
+                // should we ignore the conflict or auto commit
+                let _ = dbtx.commit_tx_result().await;
+            }
+        }
+    }
+
+    async fn update_meta_values(
+        dbtx: &mut DatabaseTransaction<'_, Committable>,
+        source_id: u64,
+        meta: &BTreeMap<MetaFieldName, MetaFieldValue>,
+    ) {
+        let stream = dbtx.find_by_prefix(&MetaFieldPrefix).await;
+
+        let to_remove = stream
+            .filter(|(key, _)| std::future::ready(key.source_id == source_id))
+            .map(|(key, _)| key)
+            .collect::<Vec<_>>()
+            .await;
+
+        for key in to_remove {
+            dbtx.remove_entry(&key).await;
+        }
+        for (name, value) in meta {
+            dbtx.insert_entry(
+                &MetaFieldKey {
+                    name: name.clone(),
+                    source_id,
+                },
+                value,
+            )
+            .await;
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+#[non_exhaustive]
+pub struct LegacyClientConfigMetaSource {}
+
+pub const LEGACY_CLIENT_CONFIG_PRIORITY: u64 = 0;
+pub const LEGACY_CLIENT_CONFIG_SOURCE_ID: u64 = 0;
+
+#[apply(async_trait_maybe_send!)]
+impl MetaSource for LegacyClientConfigMetaSource {
+    fn source_id(&self) -> u64 {
+        LEGACY_CLIENT_CONFIG_SOURCE_ID
+    }
```

Comment:
The `source_id` not just being a counter means we need to coordinate the values when creating new sources, which is slightly annoying (but that won't happen too often I hope).

Reply:
we can say the order of adding source determines the source id and priority, you have to keep it consistent and you can't remove any source in future

Reply:
Good point, or these old DB entries will never be invalidated …